spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/mail/JndiMailAutoConfigurationTests.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2012-2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.mail;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.boot.autoconfigure.jndi.JndiPropertiesHidingClassLoader;
+import org.springframework.boot.autoconfigure.jndi.TestableInitialContextFactory;
+import org.springframework.boot.test.EnvironmentTestUtils;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+import javax.mail.Session;
+import javax.naming.Context;
+import javax.naming.NamingException;
+import java.util.Properties;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test for {@link JndiMailAutoConfiguration}
+ *
+ * @author Eddú Meléndez
+ */
+public class JndiMailAutoConfigurationTests {
+
+	private ClassLoader threadContextClassLoader;
+
+	private String initialContextFactory;
+
+	private AnnotationConfigApplicationContext context;
+
+	@Before
+	public void setupJndi() {
+		this.initialContextFactory = System.getProperty(Context.INITIAL_CONTEXT_FACTORY);
+		System.setProperty(Context.INITIAL_CONTEXT_FACTORY,
+			TestableInitialContextFactory.class.getName());
+	}
+
+	@Before
+	public void setupThreadContextClassLoader() {
+		this.threadContextClassLoader = Thread.currentThread().getContextClassLoader();
+		Thread.currentThread().setContextClassLoader(
+			new JndiPropertiesHidingClassLoader(getClass().getClassLoader()));
+	}
+
+	@After
+	public void close() {
+		TestableInitialContextFactory.clearAll();
+		if (this.initialContextFactory != null) {
+			System.setProperty(Context.INITIAL_CONTEXT_FACTORY,
+				this.initialContextFactory);
+		}
+		else {
+			System.clearProperty(Context.INITIAL_CONTEXT_FACTORY);
+		}
+		if (this.context != null) {
+			this.context.close();
+		}
+		Thread.currentThread().setContextClassLoader(this.threadContextClassLoader);
+	}
+
+	@Test
+	public void mailSessionIsAvailableFromJndi() throws IllegalStateException,
+		NamingException {
+		Properties properties = new Properties();
+		Session session = Session.getDefaultInstance(properties);
+		configureJndi("foo", session);
+
+		this.context = new AnnotationConfigApplicationContext();
+		EnvironmentTestUtils.addEnvironment(this.context, "spring.mail.jndi-name:foo");
+		this.context.register(JndiMailAutoConfiguration.class);
```

Comment:
This test would benefit from actually my proposal of merging the two. You're hiding what `MailSenderAutoConfiguration` does since you're not including it. You don't care about `Session`. You care about the fact that a `JavaMailSender` is created properly. 

So you need 4 tests really

1. JNDI is present and the property is present and a `JavaMailSender` is created with the JNDI Mail `Session`
2. JNDI is present, the property is not present but the `spring.mail.host` one is so no JNDI Mail `Session` bean is created but a `JavaMailSender` based on the properties are created
3. JNDI is present, the property is not present and the `spring.mail.host` one is not either so JNDI Mail `Session` bean is created and no `JavaMailSender` either
4. JNDI is not present and the property is present. No JNDI Mail `Session` bean is created and no `JavaMailSender` either